### ISSUE TITLE
usearch_global sequences that contain repeated kmers

### ISSUE THREAD

**Issue by cnjr2** [torognes/vsearch#95]
I have been running into some unexpected behaviour and I think it is related to #85 (repeats within sequences).

My reads file `derep_reads.fasta` contains many repeats of the following short sequences:
```
>read1
TAGGTATAGGTATAGGTATAGGTATAGGTA
>read2
TAGGTATAGGTATAGGTATAGGTATAGGGA
>read3
TAGGTATGGGTATAGGTATAGGTATAGGGA
>read4
ATTATTGCTAGTTATTGCAAACTCGCTGATTCGCTATAGATCGTATGCATCGCATGATAGCTAG
...
```
My `db.fasta` contains all the centroids that I determined previously with `--cluster_size` and there seemed to be no problems:
```
>centroid1
TAGGTATGGGTATAGGTATAGGTATAGGGA
>centroid2
ATTATTGCTAGTTATTGCAAACTCGCTGATTCGCTATAGATCGTATGCATCGCATGATAGCTAG
```
I have tried to match the reads to the centroids with `usearch_global` but I do not get any hits for `>centroid1`.
```
vsearch-1.1.3-osx-x86_64 \
--usearch_global derep_reads.fasta \
--db db.fasta \
--minseqlength 1 \
--output_no_hits \
--strand plus \
--id 0.9 \
--query_cov 0.9 \
--target_cov 0.9 \
--samout mappings.sam
```
Here is also the USEARCH equivalent that worked:
```
usearch8.0.1616_i86osx32 \
-usearch_global derep_reads.fasta \
-strand plus \
-db db.fasta \
-id 0.9 \
-samout mappings.sam
```
Is there any way around this behaviour?

p.s. Thanks for the development!

**Comment by frederic-mahe** [comment]
Thanks for your report @cnjr2.

could it be an automatic soft masking? do you get the same results with `--qmask none` and `--dbmask none`?

**Comment by cnjr2** [comment]
Thank you for your help!
I have now tried soft masking the db- and query-sequences, but it seems like there are still no matches found... It does not give me any error messages or warnings of any sort, so I do not know what other information I could provide you with.

**Comment by frederic-mahe** [comment]
Then, it may be an implementation problem. Maybe Torbjørn will have an idea?

**Comment by cnjr2** [comment]
UPDATE:
The example I gave above is a simplification of the original dataset that I used, for which I reported the  unexpected behaviour. I have now run the example with exactly the sequences provided above (I should have mentioned this before - sorry). 

Contrary to before, I now get a match for `centroid1`, but only if it is a perfect match (see `read3` in the output `mappings.sam`). `read1` and `read2` are still not found despite only a single mismatch.

So for completion here is also my output `mappings.sam`: 
```
read2   4       *       0       255     *       *       0       0       TAGGTATAGGTATAGGTATAGGTATAGGGA  *
read1   4       *       0       255     *       *       0       0       TAGGTATAGGTATAGGTATAGGTATAGGTA  *
read3   0       centroid1       1       255     30M     *       0       0       TAGGTATGGGTATAGGTATAGGTATAGGGA  *       AS:i:100        XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:30 YT:Z:UU
read4   0       centroid2       1       255     64M     *       0       0       ATTATTGCTAGTTATTGCAAACTCGCTGATTCGCTATAGATCGTATGCATCGCATGATAGCTAG        *       AS:i:100        XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:64 YT:Z:UU
```

Maybe this makes things more clear...

**Comment by torognes** [comment]
Hi @cnjr2. I think you are right that the problem is related to issue #85. The first three reads all contain repeats of the sequence TAGGTA. Read 2 and 3 have some small differences. The problem is that vsearch will initially just look at the unique 8-mers in the sequences, and in this case there are very few unique 8-mers in the first 3 sequences. To find matches, at least six unique 8-mers must match between the query and any database sequence. I will look into this problem in the days ahead and hope to find a solution to the problem. Thanks for reporting the issue. - Torbjørn

**Comment by cnjr2** [comment]
Thank you.

**Comment by cnjr2** [comment]
I was wondering whether this behaviour would also influence the dereplication and the clustering?

**Comment by frederic-mahe** [comment]
hi @cnjr2, dereplication does not use k-mers, so it cannot be affected. Clustering and search use the same algorithmic core (based on k-mers), so clustering is probably affected too.

**Comment by cnjr2** [comment]
Thanks for the info!

**Comment by torognes** [comment]
Related to #85

**Comment by torognes** [comment]
This problem is solved now in version 1.4.7. This version of vsearch will consider all words in the sequences, not just those occurring exactly once.

**Comment by cnjr2** [comment]
Thank you! I will try it out and let you know how it went.
